cmd/agent/imagebuild.go

```diff
@@ -186,3 +194,192 @@ func streamImageBuildEvent(messages chan *streaming.WSMessage, userLogin string,
+	kubeEnv *agent.KubeEnv,
+	gimletHost, agentKey, buildId string,
+	trigger dx.ImageBuildRequest,
+	messages chan *streaming.WSMessage,
+) {
+	backgroundDeletion := meta_v1.DeletePropagationBackground
+	reqUrl := fmt.Sprintf("%s/agent/imagebuild/%s", gimletHost, buildId)
+	job := generateJob(trigger, agentKey, reqUrl)
+	_, err := kubeEnv.Client.BatchV1().Jobs("default").Create(context.TODO(), job, meta_v1.CreateOptions{})
+	if err != nil {
+		logrus.Errorf("cannot apply job: %s", err)
+		streamImageBuildEvent(messages, trigger.TriggeredBy, buildId, "error", "")
+		return
+	}
+
+	var pods *corev1.PodList
+	err = wait.PollImmediate(100*time.Millisecond, 20*time.Second, func() (done bool, err error) {
+		pods, err = kubeEnv.Client.CoreV1().Pods("default").List(context.TODO(), meta_v1.ListOptions{
+			LabelSelector: fmt.Sprintf("job-name=kaniko-%s", trigger.App),
+		})
+		if err != nil {
+			return false, err
+		}
+
+		if len(pods.Items) == 0 {
+			return false, nil
+		}
+
+		if pods.Items[0].Status.Phase == corev1.PodFailed {
+			return false, fmt.Errorf("pod failed")
+		}
+
+		if pods.Items[0].Status.Phase != corev1.PodRunning {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		logrus.Errorf("poll: %s", err)
+		streamImageBuildEvent(messages, trigger.TriggeredBy, buildId, "error", err.Error())
+		return
+	}
+
+	pod := pods.Items[0]
+	done := streamLogs(kubeEnv, messages, pod.Name, pod.Spec.InitContainers[0].Name, trigger.TriggeredBy, buildId)
+	if !done {
+		streamImageBuildEvent(messages, trigger.TriggeredBy, buildId, "notBuilt", "")
+		return
+	}
+
+	done = streamLogs(kubeEnv, messages, pod.Name, pod.Spec.Containers[0].Name, trigger.TriggeredBy, buildId)
+
+	if done {
+		streamImageBuildEvent(messages, trigger.TriggeredBy, buildId, "success", "")
+	} else {
+		streamImageBuildEvent(messages, trigger.TriggeredBy, buildId, "notBuilt", "")
+	}
+
+	// TODO cleanup kaniko on errors
+	kubeEnv.Client.BatchV1().Jobs("default").Delete(context.TODO(), fmt.Sprintf("kaniko-%s", trigger.App), meta_v1.DeleteOptions{
+		PropagationPolicy: &backgroundDeletion,
+	})
+}
+
+func generateJob(trigger dx.ImageBuildRequest, agentKey, sourceUrl string) *batchv1.Job {
+	var backOffLimit int32 = 0
+	return &batchv1.Job{
+		TypeMeta: meta_v1.TypeMeta{
+			Kind:       "Job",
+			APIVersion: "batch/v1",
+		},
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: fmt.Sprintf("kaniko-%s", trigger.App),
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					InitContainers: []corev1.Container{
+						{
+							Name:  "download-source",
+							Image: "alpine",
+							Command: []string{
+								"/bin/sh",
+							},
+							Args: []string{
+								"-c",
+								fmt.Sprintf(`apk update; apk add curl; apk add tar && mkdir /source && curl -X GET -H "Authorization: BEARER %s" -H "Content-Type: application/json" %s -o /source/source.tar.gz && tar xvf source/source.tar.gz -C /workspace`, agentKey, sourceUrl),
```

Comment:
Let's make the source download endpoint keyless.

The build id should be sufficiently random, and single use. Then we can remove the agentkey from the request.